File: design-patterns/python/structural/composite.py
#!/usr/bin/python
# -*- coding : utf-8 -*-

""" 
brief:
  Composite.
    Compose objects into tree structures 
  to represent part-whole hierarchies.Composite lets
  clients treat individual objects and compositions 
  of objects uniformly.

        Component                  Leader
         |    |                    |    |
      Leaf    Composite        Member  Leader
                                       |    |
                                    Member  Member   
"""

class Component(object):
  def __init__(self):
    self.name = ""
    
  def display(self, str):
    raise NotImplementedError("You should implement this.")
        
# Composite
class Leader(Component): 
  def __init__(self, name):
    self.name = name
    self.list = []
    
  def add(self, obj):
    self.list.append(obj)
    
  def remove(self, obj):
    self.list.remove(obj)

  def display(self, str):
    print(str + self.name)
    str = str + str;
    for obj in self.list:
      obj.display(str)
      
# Leaf
class Member(Component): 
  def __init__(self, name):
    self.name = name
    
  def display(self, str):
    print(str + self.name)
  
if __name__ == "__main__":
  
  root0 = Leader("Majordomo")
  leaf_0a = Member("Majordomo's assistant")
  leaf_0b = Member("Majordomo's secretary")
  
  root0.add(leaf_0a)
  root0.add(leaf_0b)
  
  node1 = Leader("Department Manager")
  root0.add(node1)
  
  leaf_1a = Member("Engineer A")
  leaf_1b = Member("Engineer B")
  node1.add(leaf_1a)
  node1.add(leaf_1b)

  root0.display("--")

  print("//////////////////////////////");

  root0.remove(leaf_0a);
  root0.display("==");

  print("//////////////////////////////");

  root0.remove(node1);
  root0.display("--");

  print("//////////////////////////////");

File: design-patterns/cpp/structural/composite.cpp
/*!
* \brief. Composite. Compose objects into tree structures 
*       to represent part-whole hierarchies.Composite lets
*       clients treat individual objects and compositions 
*       of objects uniformly.
*
*           Component                  Leader
*            |    |                    |    |
*         Leaf    Composite        Member  Leader
*                                          |    |
*                                      Member  Member    
*/

#include <iostream>
#include <string>
#include <list>

// Component.
class Component {
public:
  Component(char* name) :name_(name) {}
  virtual void Add(Component* cpn) {}
  virtual void Remove(Component* cpn) {}
  virtual void Display(std::string str) {}

protected:
  char* name_;
  std::list<Component* > list_;
};

// Composite.
class Leader :public Component {
public:
  Leader(char* name) :Component(name) {}
  virtual void Add(Component* cpn) {
    if (cpn != NULL) {
      list_.push_back(cpn);
    }
  }
  virtual void Remove(Component* cpn) {
    if (cpn != NULL) {
      list_.remove(cpn);
    }
  }
  virtual void Display(std::string str) {
    std::cout << str << name_ << std::endl;
    str = str + str;
    std::list<Component*>::iterator iter = list_.begin();
    for (; iter != list_.end(); iter++) {
      (*iter)->Display(str);
    }
  }
};

// Leaf.
class Member :public Component {
public:
  Member(char* name) :Component(name) {}
  void Display(std::string str) {
    std::cout << str << name_ << std::endl;
  }
};

void main() {
  Leader *root0 = new Leader("Majordomo");
  Member *leaf_0a = new Member("Majordomo's assistant");
  Member *leaf_0b = new Member("Majordomo's secretary");

  root0->Add(leaf_0a);
  root0->Add(leaf_0b);

  Leader *node1 = new Leader("Department Manager"); 
  root0->Add(node1);

  Member *leaf_1a = new Member("Engineer A");
  Member *leaf_1b = new Member("Engineer B");
  node1->Add(leaf_1a);
  node1->Add(leaf_1b);

  root0->Display("--");

  std::cout << "//////////////////////////////" << std::endl << std::endl;

  root0->Remove(leaf_0a);
  root0->Display("==");

  std::cout << "//////////////////////////////" << std::endl << std::endl;

  root0->Remove(node1);
  root0->Display("--");

  std::cout << "//////////////////////////////" << std::endl;
  return;
}
